man/Cart3Convert.1
.\" -*- mode: troff; coding: utf-8 -*-
.\" Automatically generated by Pod::Man v6.0.2 (Pod::Simple 3.45)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" \*(C` and \*(C' are quotes in nroff, nothing in troff, for use with C<>.
.ie n \{\
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Required to disable full justification in groff 1.23.0.
.if n .ds AD l
.\" ========================================================================
.\"
.IX Title "CART3CONVERT 1"
.TH CART3CONVERT 1 2025-11-05 "GeographicLib 2.7" "GeographicLib Utilities"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH NAME
Cart3Convert \-\- convert geocentric coordinates for a triaxial ellipsoid
.SH SYNOPSIS
.IX Header "SYNOPSIS"
\&\fBCart3Convert\fR [ \fB\-E\fR | \fB\-G\fR | \fB\-P\fR | \fB\-C\fR | \fB\-GX\fR | \fB\-PX\fR | \fB\-CX\fR ]
[ \fB\-r\fR ] [ \fB\-3\fR | \fB\-D\fR ]
[ \fB\-R\fR \fInrand\fR ] [ \fB\-\-seed\fR \fIs\fR ]
[ \fB\-t\fR \fIa\fR \fIb\fR \fIc\fR | \fB\-e\fR \fIb\fR \fIe2\fR \fIk2\fR \fIkp2\fR ]
[ \fB\-d\fR | \fB\-:\fR ] [ \fB\-w\fR ] [ \fB\-p\fR \fIprec\fR ]
[ \fB\-\-comment\-delimiter\fR \fIcommentdelim\fR ]
[ \fB\-\-version\fR | \fB\-h\fR | \fB\-\-help\fR ]
[ \fB\-\-input\-file\fR \fIinfile\fR | \fB\-\-input\-string\fR \fIinstring\fR ]
[ \fB\-\-line\-separator\fR \fIlinesep\fR ]
[ \fB\-\-output\-file\fR \fIoutfile\fR ]
.SH DESCRIPTION
.IX Header "DESCRIPTION"
Convert coordinates for a triaxial ellipsoid to geocentric coordinates.
Geocentric coordinates have the origin at the center of the earth, with
the \fIX\fR, \fIY\fR, and \fIZ\fR axes aligned with the major, median, and minor
semiaxes, \fIa\fR, \fIb\fR, and \fIc\fR.  By default, the conversion is to
geocentric coordinates.  Specifying \fB\-r\fR causes the reverse conversion
to be performed.
.PP
Seven latitude/longitude/heading coordinate systems are supported
.IP 1. 4
Ellipsoidal coordinates, \fIbeta\fR, \fIomega\fR, and \fIalpha\fR, signified by
\&\fB\-E\fRn (the default)
.IP 2. 4
Geodetic coordinates, \fIphi\fR, \fIlambda\fR, and \fIzeta\fR, signified by
\&\fB\-G\fR.
.IP 3. 4
Parametric coordinates, \fIphi\*(Aq\fR, \fIlambda\*(Aq\fR, and \fIzeta\fR, signified by
\&\fB\-P\fR.
.IP 4. 4
Geocentric coordinates, \fIphi"\fR, \fIlambda"\fR, and \fIzeta\fR, signified by
\&\fB\-C\fR.
.IP 5. 4
Geodetic coordinates, \fIphi*\fR, \fIlambda*\fR, and \fIzeta*\fR, with the north
pole lying on the \fIX\fR axis, signified by \fB\-GX\fR.
.IP 6. 4
Parametric coordinates, \fIphi*\*(Aq\fR, \fIlambda*\*(Aq\fR, and \fIzeta*\fR, with the
north pole lying on the \fIX\fR axis, signified by \fB\-PX\fR.
.IP 7. 4
Geocentric coordinates, \fIphi*"\fR, \fIlambda*"\fR, and \fIzeta*\fR, with the
north pole lying on the \fIX\fR axis, signified by \fB\-CX\fR.
.PP
In general, only ellipsoidal coordinates are orthogonal.  However for
all coordinates, we define the heading as the direction measured
clockwise from a line of constant longitude.  \fIzeta\fR is the same for
the geodetic, parametric, and geocentric coordinate systems.
.PP
Points away from the surface of the ellipsoid can be described by the
orthogonal height \fIh\fR above from the ellipsoid and the position is
given by \fIphi\fR, \fIlambda\fR (giving the closest point on the ellipsoid),
and \fIh\fR.  Alternatively, the confocal ellipsoid of minor radius \fIc\fR +
\&\fIH\fR point can be found and the position of the point is then given by
\&\fIbeta\fR, \fIomega\fR for the confocal ellipsoid and \fIH\fR.
.PP
By default, coordinates are provided on standard input as a set of lines
containing (blank separated) \fIlatitude\fR and \fIlongitude\fR (decimal
degrees or degrees, minutes and seconds).  The interpretation of the
coordinates is governed by the \fB\-E\fR, \fB\-G\fR, \fB\-P\fR, \fB\-C\fR \fB\-GX\fR,
\&\fB\-PX\fR, and \fB\-CX\fR flags.  For each set of coordinates, the
corresponding cartesian coordinates \fIX\fR, \fIY\fR, \fIZ\fR (meters) are
printed on standard output.  If \fI\-r\fR is provided, the reserve operation
is undertaken reading \fIX\fR, \fIY\fR, \fIZ\fR (which are assumed to denote a
point on the ellipsoid) on standard input and printing the \fIlatitude\fR
and \fIlongitude\fR.
.PP
If the \fB\-3\fR flag is given (this cannot be combined with \fB\-D\fR), an
arbitrary point in three\-dimensional space is specified either with
\&\fIbeta\fR \fIomega\fR \fIH\fR or with \fIphi\fR \fIlambda\fR \fIh\fR.
.PP
If the \fB\-D\fR flag is given (this cannot be combined with \fB\-3\fR), the
heading (direction) is included in the computation converting \fIbeta\fR
\&\fIomega\fR \fIalpha\fR or \fIphi\fR \fIlambda\fR \fIzeta\fR to \fIX\fR, \fIY\fR, \fIZ\fR \fIVx\fR,
\&\fIVy\fR, \fIVz\fR where [\fIvx\fR, \fIvy\fR, \fIvz\fR] is a unit vector lying in the
surface of the ellipsoid.
.PP
If the \fB\-R\fR flag is given, do not read any input; instead output random
cartesian points lying on the surface of the ellipsoid.  If the \fB\-r\fR
flag is specified the cartesian points are converted into one of the
triaxial coordinate systems.
.SH OPTIONS
.IX Header "OPTIONS"
.IP \fB\-E\fR 4
.IX Item "-E"
\&\fIlatitude\fR and \fIlongitude\fR is interpreted as ellipsoidal coordinates
\&\fIbeta\fR and \fIomega\fR.  This is the default.
.IP "\fB\-G\fR or \fB\-GX\fR" 4
.IX Item "-G or -GX"
\&\fIlatitude\fR and \fIlongitude\fR is interpreted as geodetic coordinates
\&\fIphi\fR and \fIlambda\fR.  The \fBX\fR added to the flags means that the north
pole lies on the \fIX\fR axis.
.IP "\fB\-P\fR or \fB\-PX\fR" 4
.IX Item "-P or -PX"
\&\fIlatitude\fR and \fIlongitude\fR is interpreted as parametric coordinates
\&\fIphi\*(Aq\fR and \fIlambda\*(Aq\fR.  The \fBX\fR added to the flags means that the north
pole lies on the \fIX\fR axis.
.IP "\fB\-C\fR or \fB\-CX\fR" 4
.IX Item "-C or -CX"
\&\fIlatitude\fR and \fIlongitude\fR is interpreted as geocentric coordinates
\&\fIphi"\fR and \fIlambda"\fR.  The \fBX\fR added to the flags means that the north
pole lies on the \fIX\fR axis.
.IP \fB\-r\fR 4
.IX Item "-r"
perform the reverse projection from \fIx\fR, \fIy\fR, \fIz\fR to \fIlatitude\fR,
\&\fIlongitude\fR.  Unless \fB\-3\fR is given, it is assumed that the point lies
on the surface of the ellipsoid.
.IP \fB\-3\fR 4
.IX Item "-3"
include the height \fIh\fR or \fIH\fR in the calculation.  This cannot be
combined with \fB\-D\fR.
.IP \fB\-D\fR 4
.IX Item "-D"
include the direction \fIalpha\fR or \fIzeta\fR in the calculation.  With the
\&\fB\-r\fR option, it is assumed that the vector [\fIvx\fR, \fIvy\fR, \fIvz\fR] is
tangent to the surface of the ellipsoid.  This cannot be combined with
\&\fB\-3\fR.
.IP "\fB\-R\fR \fInrand\fR" 4
.IX Item "-R nrand"
output \fInrand\fR random points on the surface of the ellipsoid as
cartesian coordinates.  With \fB\-D\fR add also a random cartesian direction
tangent to the surface of the ellipsoid.  In this mode, nothing is read
from standard input.  With the \fB\-r\fR option the random points are
converted into one of the triaxial coordinate systems (and the direction
is converted to a heading if \fB\-D\fR is given).
.IP "\fB\-\-seed\fR \fIs\fR" 4
.IX Item "--seed s"
set the random number seed to \fIs\fR when using \fB\-R\fR.  \fIs\fR = 0 (the
default) causes a random seed to be used.
.IP "\fB\-t\fR \fIa\fR \fIb\fR \fIc\fR" 4
.IX Item "-t a b c"
specify the ellipsoid via its major semiaxis \fIa\fR, median semiaxis \fIb\fR,
and minor semixis \fIc\fR.  By default, we have \fIa\fR = 6378172 m, \fIb\fR =
6378102 m, \fIc\fR = 6356752 m, an approximate triaxial model of the earth.
(With this model \fIomg\fR = 0deg, corresponds to \fIlon\fR = \-14.93deg.
.IP "\fB\-e\fR \fIb\fR \fIe2\fR \fIk2\fR \fIkp2\fR" 4
.IX Item "-e b e2 k2 kp2"
specify the ellipsoid via the median semiaxis, \fIb\fR and the shape
parameters \fIe2\fR = (\fIa\fR^2 \- \fIc\fR^2)/\fIb\fR^2, \fIk2\fR = (\fIb\fR^2 \-
\&\fIc\fR^2)/(\fIa\fR^2 \- \fIc\fR^2), and \fIkp2\fR = (\fIa\fR^2 \- \fIb\fR^2)/(\fIa\fR^2 \-
\&\fIc\fR^2).  Simple fractions are allowed for \fIe2\fR, \fIk2\fR, and \fIkp2\fR.
Internally, the supplied values of \fIk2\fR and \fIkp2\fR are normalized so
that \fIk2\fR + \fIkp2\fR = 1.
.IP \fB\-d\fR 4
.IX Item "-d"
output angles as degrees, minutes, seconds instead of decimal degrees.
.IP \fB\-:\fR 4
.IX Item "-:"
like \fB\-d\fR, except use : as a separator instead of the d, \*(Aq, and "
delimiters.
.IP \fB\-w\fR 4
.IX Item "-w"
toggle the longitude first flag (it starts off); if the flag is on, then
on input and output, longitude precedes latitude (except that, on input,
this can be overridden by a hemisphere designator, \fIN\fR, \fIS\fR, \fIE\fR,
\&\fIW\fR).
.IP "\fB\-p\fR \fIprec\fR" 4
.IX Item "-p prec"
set the output precision to \fIprec\fR (default 3).  For distances, \fIprec\fR
is the number of digits after the decimal point for ellipsoids which are
approximately the same size as the Earth; for other ellipsoids the
precision is adjusted to retain the same relative precision.  For
latitudes and longitudes (in degrees), the number of digits after the
decimal point is \fIprec\fR + 5.  For cartesian directions, the precision is
\&\fIprec\fR + 7.
.IP "\fB\-\-comment\-delimiter\fR \fIcommentdelim\fR" 4
.IX Item "--comment-delimiter commentdelim"
set the comment delimiter to \fIcommentdelim\fR (e.g., "#" or "//").  If
set, the input lines will be scanned for this delimiter and, if found,
the delimiter and the rest of the line will be removed prior to
processing and subsequently appended to the output line (separated by a
space).
.IP \fB\-\-version\fR 4
.IX Item "--version"
print version and exit.
.IP \fB\-h\fR 4
.IX Item "-h"
print usage and exit.
.IP \fB\-\-help\fR 4
.IX Item "--help"
print full documentation and exit.
.IP "\fB\-\-input\-file\fR \fIinfile\fR" 4
.IX Item "--input-file infile"
read input from the file \fIinfile\fR instead of from standard input; a file
name of "\-" stands for standard input.
.IP "\fB\-\-input\-string\fR \fIinstring\fR" 4
.IX Item "--input-string instring"
read input from the string \fIinstring\fR instead of from standard input.
All occurrences of the line separator character (default is a semicolon)
in \fIinstring\fR are converted to newlines before the reading begins.
.IP "\fB\-\-line\-separator\fR \fIlinesep\fR" 4
.IX Item "--line-separator linesep"
set the line separator character to \fIlinesep\fR.  By default this is a
semicolon.
.IP "\fB\-\-output\-file\fR \fIoutfile\fR" 4
.IX Item "--output-file outfile"
write output to the file \fIoutfile\fR instead of to standard output; a
file name of "\-" stands for standard output.
.SH EXAMPLES
.IX Header "EXAMPLES"
.Vb 4
\&   echo 1 2 3 | Cart3Convert \-t 3 2 1 \-3 \-p 0 \-r
\&   => 58.69140 75.11263 2.5860649
\&   echo 58.69140 75.11263 2.5860649 | Cart3Convert \-t 3 2 1 \-3 \-p 0
\&   => 1.0000001 2.0000003 2.9999999
.Ve
.SH ERRORS
.IX Header "ERRORS"
An illegal line of input will print an error message to standard output
beginning with \f(CW\*(C`ERROR:\*(C'\fR and causes \fBCart3Convert\fR to return an exit
code of 1.  However, an error does not cause \fBCart3Convert\fR to
terminate; following lines will be converted.
.SH AUTHOR
.IX Header "AUTHOR"
\&\fBCart3Convert\fR was written by Charles Karney.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fBGeod3Solve\fR\|(1).
.PP
The algorithms are described in Appendices A and B of
C. F. F. Karney,
\&\fIJacobi\*(Aqs solution for geodesics on a triaxial ellipsoid\fR,
Technical Report, SRI International, Nov. 2025,
<https://arxiv.org/abs/2511.01621>.
.PP
The Wikipedia page, Geodesics on an ellipsoid,
<https://en.wikipedia.org/wiki/Geodesics_on_an_ellipsoid>.
.SH HISTORY
.IX Header "HISTORY"
\&\fBCart3Convert\fR was added to GeographicLib,
<https://geographiclib.sourceforge.io>, in version 2.6.


NEWS
A reverse chronological list of changes to GeographicLib

For more information, see

    https://geographiclib.sourceforge.io/C++/doc/

The current version of the library is 2.7.

Changes between 2.7 (released 2025-11-06) and 2.6 versions:

   * Fix two bugs in classes for triaxial ellipsoids:
     + Triaxial::Geodesic3::Inverse sometimes threw an exception when
       both endpoints were on the major principal ellipse for very
       eccentric ellipsoids a > 2 c.
     + Triaxial::Cartesian3::carttocart2 sometimes returned the wrong
       result in the interior of the ellipsoid with Z = 0.

   * Cite the paper covering some of the triaxial classes Jacobi's
     solution for geodesics on a triaxial ellipsoid.

   * Minor documentation fixes.

Changes between 2.6 (released 2025-09-30) and 2.5.2 versions:

   * Support for triaxial ellipsoids. This is all placed in the
     GeographicLib::Triaxial namespace with the header files in the
     GeographicLib/Triaxial directory. This provides
     + Triaxial::Ellipsoid3 supports basic conversions of ellipsoidal
       coordinates.
     + Triaxial::GeodesicLine3 and Triaxial::Geodesic3 solve the direct
       and inverse geodesic problems on a triaxial ellipsoid. This
       implements Jacobi's solution for the direct problem. The inverse
       problem extends the method for solving the inverse problem on a
       biaxial ellipsoid.
     + Triaxial::Cartesian3 performs conversions between different
       systems of triaxial coordinates.
     + Triaxial::Conformal3 implements Jacobi's conformal projection for
       a triaxial ellipsoid. This allow a triaxial ellipsoid to be
       conformally mapped to a sphere or to any other ellipsoid.
     + Geod3Solve is a command-line utility offering access to
       Triaxial::Geodesic3.
     + Cart3Convert is a command-line utility offering access to
       Triaxial::Cartesian3.
     + Conformal3Proj is a command-line utility offering access to
       Triaxial::Conformal3.
     + Constants::Triaxial_Earth_a, Constants::Triaxial_Earth_b,
       Constants::Triaxial_Earth_c, and Constants::Triaxial_Earth_lon0
       specify a simple triaxial model for the Earth.
     + experimental::TriaxialGeodesicODE is provided to illustrate the
       solution of the direct geodesic problem by integrating the
       ordinary differential equations for the geodesic. Because this
       depends on Boost, this is not part of the GeographicLib library.
     + experimental/Geod3ODE.cpp a utility for doing direct geodesic
       calculations using experimental::TriaxialGeodesicODE.

   * To support the triaxial routines, the following classes have been
     added:
     + AngleT provides an accurate representation of angles of any
       magnitude. This might replace AuxAngle at some point.
     + Trigfun supports representing periodic functions as Fourier
       series with automatic selection of the number of terms to
       use. This offers a tiny subset of the functionality of the trig
       series in Chebfun. This might replace DST at some point.

   * C++17 is now required. Minimum version of Visual Studio supported
     is Visual Studio 15 2017.

   * Add Math::clamp.

   * Because experimental::TriaxialGeodesicODE depends on Boost, the
     CMake option USE_BOOST_FOR_EXAMPLES has been renamed
     USE_BOOST. This will cause find_package(Boost) to be called, but
     does not affect how the library or the tools are built.

Changes between 2.5.2 (released 2025-08-22) and 2.5.1 versions:

   * Fix potential buffer overflow in DMS.  Upgrading from earlier
     versions is strongly recommended.  If upgrading is not possible,
     avoid processing untrusted or malformed inputs and run programs
     linked to the library with restricted privileges.  Thanks to
     Rosario Matteo Grammatico for finding this vulnerability.

Changes between 2.5.1 (released 2025-08-20) and 2.5 versions:

   * Fix bug in geodesic inverse problem with very prolate
     ellipsoids. For f < -2, a meridional path was incorrectly
     returned in some cases for two points on the same meridian. This
     affected both the series and exact solutions. (Or course, the
     series solution is inaccurate for ellipsoids this eccentric.)

   * Minor tweak to Math::atan2d to improve accuracy.

   * Allow GEOGRAPHICLIB_PRECISION > 5 to allow use of mpreal with a
     fixed precision.

   * Document that cmake option -D BINDIR=OFF removes tools from the
     default all target for make.

   * WARNING: C++17 compliant compiler will be required for the next
     release. Support for builds with Visual Studio 14 (2015) will
     then stop.

Changes between 2.5 (released 2024-12-28) and 2.4 versions:

   * New magnetic models provided

     + World Magnetic Model 2025, wmm2025; this covers the period
       2025-2030.  This is now the model returned by
       MagneticModel::DefaultMagneticName and is default magnetic
       model for MagneticField (replacing wmm2020 which is only valid
       thru the end of 2024).
     + World Magnetic Model High Resolution 2025, wmmhr2025; this is a
       high-resolution version of wmm2025 including terms up to degree
       133 (as opposed to 12 for wmm2025).
     + International Geomagnetic Reference Field (14th generation),
       igrf14, which approximates the main magnetic field of the earth
       for the period 1900-2030.

   * Update CMake's cmake_minimum_required to 3.17.0 (to facilitate
     integration with Qt projects).

   * For GEOGRAPHICLIB_PRECISION > 2, readarray treats doubles as
     approximations to exact decimal constants.

Changes between 2.4 (released 2024-07-14) and 2.3 versions:

   * Math::sind, Math::sincosd, etc., now return accurate results for
     arguments = 30d, 45d, 60d, ...

   * Add Math::hypot3.

   * Add EllipticFunction::am.

   * Make constructor for AuxAngle explicit.

   * Include notes for plate carree areas.

   * C++14 compliant compiler required (this requirement tracks the

tools/Cart3Convert.cpp
/**
 * \file Cart3Convert.cpp
 * \brief Command line utility for computing geodesics on a triaxial ellipsoid
 *
 * Copyright (c) Charles Karney (2024-2025) <karney@alum.mit.edu> and licensed
 * under the MIT/X11 License.  For more information, see
 * https://geographiclib.sourceforge.io/
 *
 * See the <a href="Cart3Convert.1.html">man page</a> for usage information.
 **********************************************************************/

// Counterpart of CartConvert

// Usual flags
//   -r (x, y, z to ellipsoidal to geographic)
//   -e (supplement with -t)
//   -w longfirst
//   -p prec
//   -h help
//   -d dms
//   -: colon
//   --help
//   --version
//   --comment-delimiter
//   + other I/O related flags

// New flags
//   -E ellipsoidal coords (default)
//   -G geodetic
//   -P parametric
//   -C geocentric
//   -3 include height (only for -E and -G)
//   -D include direction (only for -E without -3)

// Can't specify -3 and -D together
// Height for -G has its usual definition
// Height for -E is the increase in the minor semiaxis for the confocal
// ellipsoid (H = u - c).

#include <iostream>
#include <iomanip>
#include <string>
#include <sstream>
#include <fstream>
#include <GeographicLib/DMS.hpp>
#include <GeographicLib/Utility.hpp>
#include <GeographicLib/Angle.hpp>
#include <GeographicLib/Triaxial/Cartesian3.hpp>

#include "Cart3Convert.usage"

int main(int argc, const char* const argv[]) {
  try {
    using namespace GeographicLib;
    using namespace Triaxial;
    Utility::set_digits();
    using real = Math::real;
    using ang = Angle;
    using vec3 = Ellipsoid3::vec3;
    using coord = Cartesian3::coord;
    enum { GEODETIC, PARAMETRIC, GEOCENTRIC, ELLIPSOIDAL };
    coord mode = coord::ELLIPSOIDAL;
    bool threed = false, direction = false, reverse = false, dms = false,
      longfirst = false, randompts = false;
    real
      a = Constants::Triaxial_Earth_a(),
      b = Constants::Triaxial_Earth_b(),
      c = Constants::Triaxial_Earth_c(),
      e2 = -1, k2 = 0, kp2 = 0;
    int prec = 3, nrand = 0;
    unsigned long long seed = 0;
    std::string istring, ifile, ofile, cdelim;
    char lsep = ';', dmssep = char(0);

    for (int m = 1; m < argc; ++m) {
      std::string arg(argv[m]);
      if (arg == "-E")
        mode = coord::ELLIPSOIDAL;
      else if (arg == "-G")
        mode = coord::GEODETIC;
      else if (arg == "-P")
        mode = coord::PARAMETRIC;
      else if (arg == "-C")
        mode = coord::GEOCENTRIC;
      else if (arg == "-GX")
        mode = coord::GEODETIC_X;
      else if (arg == "-PX")
        mode = coord::PARAMETRIC_X;
      else if (arg == "-CX")
        mode = coord::GEOCENTRIC_X;
      else if (arg == "-R") {
        if (++m == argc) return usage(1, true);
        try {
          nrand = Utility::val<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Number of randoms " << argv[m] << " is not a number\n";
          return 1;
        }
        randompts = true;
      } else if (arg == "-3")
        threed = true;
      else if (arg == "-D")
        direction = true;
      else if (arg == "-r")
        reverse = true;
      else if (arg == "--seed") {
        if (++m == argc) return usage(1, true);
        try {
          seed = Utility::val<unsigned long long>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "-t") {
        if (m + 3 >= argc) return usage(1, true);
        try {
          a = Utility::val<real>(std::string(argv[m + 1]));
          b = Utility::val<real>(std::string(argv[m + 2]));
          c = Utility::val<real>(std::string(argv[m + 3]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -t: " << e.what() << "\n";
          return 1;
        }
        e2 = -1;
        m += 3;
      } else if (arg == "-e") {
        // Cayley ellipsoid sqrt([2,1,1/2]) is
        // -e 1 3/2 1/3 2/3
        if (m + 4 >= argc) return usage(1, true);
        try {
          b = Utility::val<real>(std::string(argv[m + 1]));
          e2 = Utility::fract<real>(std::string(argv[m + 2]));
          k2 = Utility::fract<real>(std::string(argv[m + 3]));
          kp2 = Utility::fract<real>(std::string(argv[m + 4]));
        }
        catch (const std::exception& e) {
          std::cerr << "Error decoding arguments of -e: " << e.what() << "\n";
          return 1;
        }
        a = -1;
        m += 4;
      } else if (arg == "-d") {
        dms = true;
        dmssep = '\0';
      } else if (arg == "-:") {
        dms = true;
        dmssep = ':';
      } else if (arg == "-w")
        longfirst = !longfirst;
      else if (arg == "-p") {
        if (++m == argc) return usage(1, true);
        try {
          prec = Utility::val<int>(std::string(argv[m]));
        }
        catch (const std::exception&) {
          std::cerr << "Precision " << argv[m] << " is not a number\n";
          return 1;
        }
      } else if (arg == "--input-string") {
        if (++m == argc) return usage(1, true);
        istring = argv[m];
      } else if (arg == "--input-file") {
        if (++m == argc) return usage(1, true);
        ifile = argv[m];
      } else if (arg == "--output-file") {
        if (++m == argc) return usage(1, true);
        ofile = argv[m];
      } else if (arg == "--line-separator") {
        if (++m == argc) return usage(1, true);
        if (std::string(argv[m]).size() != 1) {
          std::cerr << "Line separator must be a single character\n";
          return 1;
        }
        lsep = argv[m][0];
      } else if (arg == "--comment-delimiter") {
        if (++m == argc) return usage(1, true);
        cdelim = argv[m];
      } else if (arg == "--version") {
        std::cout << argv[0] << ": GeographicLib version "
                  << GEOGRAPHICLIB_VERSION_STRING << "\n";
        return 0;
      } else
        return usage(!(arg == "-h" || arg == "--help"), arg != "--help");
    }

    Cartesian3 tc(e2 >= 0 ? Ellipsoid3(b, e2, k2, kp2) : Ellipsoid3(a, b, c));

    if (randompts) {
      if (threed) {
        std::cerr << "Cannot specify -3 for random points\n";
        return 1;
      }
      if (!ifile.empty() || !istring.empty()) {
        std::cerr
          << "Cannot specify --input-{string,file} with random points\n";
      }
    } else {
      if (direction && threed) {
        std::cerr << "Cannot specify both -3 and -D\n";
        return 1;
      }
    }

    if (!ifile.empty() && !istring.empty()) {
      std::cerr << "Cannot specify --input-string and --input-file together\n";
      return 1;
    }
    if (ifile == "-") ifile.clear();
    std::ifstream infile;
    std::istringstream instring;
    if (!ifile.empty()) {
      infile.open(ifile.c_str());
      if (!infile.is_open()) {
        std::cerr << "Cannot open " << ifile << " for reading\n";
        return 1;
      }
    } else if (!istring.empty()) {
      std::string::size_type m = 0;
      while (true) {
        m = istring.find(lsep, m);
        if (m == std::string::npos)
          break;
        istring[m] = '\n';
      }
      instring.str(istring);
    }
    std::istream* input = !ifile.empty() ? &infile :
      (!istring.empty() ? &instring : &std::cin);

    std::ofstream outfile;
    if (ofile == "-") ofile.clear();
    if (!ofile.empty()) {
      outfile.open(ofile.c_str());
      if (!outfile.is_open()) {
        std::cerr << "Cannot open " << ofile << " for writing\n";
        return 1;
      }
    }
    std::ostream* output = !ofile.empty() ? &outfile : &std::cout;

    // Max precision = 10: 0.1 nm in distance, 10^-15 deg (= 0.11 nm),
    // 10^-11 sec (= 0.3 nm).
    prec = std::min(10 + Math::extra_digits(), std::max(0, prec));
    using std::round, std::log10, std::ceil, std::signbit;
    int disprec = std::max(0, prec + int(round(log10(6400000/b)))),
      angprec = prec + 5, vecprec = prec + 7;
    if (randompts) {
      //                         C a r t 3 C o n
      unsigned long long s1 = 0x4361727433436f6eULL, s2 = seed;
      if (seed == 0) {
        s1 = std::random_device()();
        s2 = std::random_device()();
      }
      std::seed_seq seq{s1, s2};
      std::mt19937 g(seq);
      for (int i = 0; i < nrand; ++i) {
        vec3 r, v = {0,0,0};
        if (direction)
          tc.cart2rand(g, r, v);
        else
          tc.cart2rand(g, r);
        if (!reverse) {
          *output << Utility::str(r[0], disprec) << " "
                  << Utility::str(r[1], disprec) << " "
                  << Utility::str(r[2], disprec);
          if (direction)
            *output << " "
                    << Utility::str(v[0], vecprec) << " "
                    << Utility::str(v[1], vecprec) << " "
                    << Utility::str(v[2], vecprec);
        } else {
          ang bet, omg, alp;
          tc.cart2toany(r, v, mode, bet, omg, alp);
          *output << ang::LatLonString(bet, omg,
                                       angprec, dms, dmssep, longfirst);
          if (direction)
            *output << " " << ang::AzimuthString(alp, angprec, dms, dmssep);
        }
        *output << "\n";
      }
      return 0;
    }
    std::string s, eol, sbet, somg, salp, sh, sx, sy, sz, svx, svy, svz, strc;
    std::istringstream str;
    int retval = 0;
    vec3 r = {0,0,0}, v = {0,0,0};
    ang bet, omg, alp;
    real h = 0;
    while (std::getline(*input, s)) {
      try {
        eol = "\n";
        if (!cdelim.empty()) {
          std::string::size_type m = s.find(cdelim);
          if (m != std::string::npos) {
            eol = " " + s.substr(m) + "\n";
            s = s.substr(0, m);
          }
        }
        // READ
        str.clear(); str.str(s);
        if (reverse) {
          if (!(str >> sx >> sy >> sz))
            throw GeographicErr("Incomplete input: " + s);
          r[0] = Utility::val<real>(sx);
          r[1] = Utility::val<real>(sy);
          r[2] = Utility::val<real>(sz);
          if (direction) {
            if (!(str >> svx >> svy >> svz))
              throw GeographicErr("Incomplete input: " + s);
            v[0] = Utility::val<real>(svx);
            v[1] = Utility::val<real>(svy);
            v[2] = Utility::val<real>(svz);
          }
        } else {
          if (!(str >> sbet >> somg))
            throw GeographicErr("Incomplete input: " + s);
          ang::DecodeLatLon(sbet, somg, bet, omg, longfirst);
          if (mode != coord::ELLIPSOIDAL && (bet.n() != 0 || signbit(bet.c())))
            throw GeographicErr("Latitude outside range [-90,90]: " + s);
          if (threed) {
            if (!(str >> sh))
              throw GeographicErr("Incomplete input: " + s);
            h = Utility::val<real>(sh);
          } else if (direction) {
            if (!(str >> salp))
              throw GeographicErr("Incomplete input: " + s);
            alp = ang::DecodeAzimuth(salp);
          }
        }
        if (str >> strc)
          throw GeographicErr("Extraneous input: " + strc);
        // PROCESS
        if (reverse) {
          if (threed)
            tc.carttoany(r, mode, bet, omg, h);
          else if (direction)
            tc.cart2toany(r, v, mode, bet, omg, alp);
          else
            tc.cart2toany(r, mode, bet, omg);
        } else {
          if (threed)
            tc.anytocart(mode, bet, omg, h, r);
          else if (direction)
            tc.anytocart2(mode, bet, omg, alp, r, v);
          else
            tc.anytocart2(mode, bet, omg, r);
        }
        // WRITE
        if (reverse) {
          *output << ang::LatLonString(bet, omg,
                                       angprec, dms, dmssep, longfirst);
          if (threed)
            *output << " " << Utility::str(h, disprec);
          else if (direction)
            *output << " " << ang::AzimuthString(alp, angprec, dms, dmssep);
        } else {
          *output << Utility::str(r[0], disprec) << " "
                  << Utility::str(r[1], disprec) << " "
                  << Utility::str(r[2], disprec);
          if (direction)
            *output << " "
                    << Utility::str(v[0], vecprec) << " "
                    << Utility::str(v[1], vecprec) << " "
                    << Utility::str(v[2], vecprec);
        }
        *output << eol;
      }
      catch (const std::exception& e) {
        // Write error message cout so output lines match input lines
        *output << "ERROR: " << e.what() << "\n";
        retval = 1;
      }
    }
    return retval;
  }
  catch (const std::exception& e) {
    std::cerr << "Caught exception: " << e.what() << "\n";
    return 1;
  }
  catch (...) {
    std::cerr << "Caught unknown exception\n";
    return 1;
  }
}
